cmd/juju/machine/add.go

```diff
@@ -170,6 +171,44 @@ type MachineManagerAPI interface {
 
 var manualProvisioner = manual.ProvisionMachine
 
+// splitUserHost given a host string of example user@192.168.122.122
+// it will return user and 192.168.122.122
+func splitUserHost(host string) (string, string) {
+	if at := strings.Index(host, "@"); at != -1 {
+		return host[:at], host[at+1:]
+	}
+	return "", host
+}
+
+func (c *addCommand) scopeRun(client AddMachineAPI, config *config.Config, ctx *cmd.Context) error {
+	logger.Infof("manual provisioning")
+
+	authKeys, err := common.ReadAuthorizedKeys(ctx, "")
+	if err != nil {
+		return errors.Annotate(err, "reading authorized-keys")
+	}
+
+	user, host := splitUserHost(c.Placement.Directive)
+	args := manualcommon.ProvisionMachineArgs{
```

Comment:
Right, but you don't have to expose manualcommon like that. One option would be to have package manual define a ProvisionMachineArgs struct, and then define it *again* inside package manual/common. The implementations would point at the manual/common one.

I think what I'd prefer, though, is to not have the manual package depend on the implementations at all. Instead, have something like:

```
manual/
    sshprovisioner/
    winrmprovisioner/
    ...
```

with the `*provisioner` packages depending on environs/manual for the common interface. Inside environs/manual, you would have an interface that each of those provisioners implements, something like:

```
type ManualProvisioner interface {
    ProvisionMachine(ProvisionMachineParams) (machineId string, err error)
}
```

And then you would have a registry of ManualProvisioners that you can get one by passing in a placement directive. AFAICR, we only need that registry at the CLI, so it could be defined there. Preferably not a global.